libs/isograph-react/src/loadable-hooks/useSkipLimitPagination.ts

```diff
@@ -75,44 +79,74 @@ export function useSkipLimitPagination<
-  function readCompletedFragmentReferences(
+  function subscribeCompletedFragmentReferences(
     completedReferences: ReadonlyArray<
       ItemCleanupPair<ReferenceCountedPointer<ArrayFragmentReference<TItem>>>
     >,
   ) {
     // In general, this will not suspend. But it could, if there is missing data.
     // A better version of this hook would not do any reading here.
-    const results = completedReferences.map(([pointer]) => {
-      const fragmentReference = pointer.getItemIfNotDisposed();
-      if (fragmentReference == null) {
-        throw new Error(
-          'FragmentReference is unexpectedly disposed. \
+    const results = completedReferences.map(
+      (
+        [pointer],
+        i,
+      ): {
+        records: WithEncounteredRecords<TItem>;
+        callback: (updatedRecords: WithEncounteredRecords<TItem>) => void;
+        fragmentReference: FragmentReference<TItem, any>;
+        readerAst: ReaderAst<TItem>;
+      } => {
+        const fragmentReference = pointer.getItemIfNotDisposed();
+        if (fragmentReference == null) {
+          throw new Error(
+            'FragmentReference is unexpectedly disposed. \
           This is indicative of a bug in Isograph.',
+          );
+        }
+
+        maybeUnwrapNetworkRequest(
+          fragmentReference.networkRequest,
+          networkRequestOptions,
+        );
+        const data = readButDoNotEvaluate(
+          environment,
+          fragmentReference,
+          networkRequestOptions,
         );
-      }
 
-      maybeUnwrapNetworkRequest(
-        fragmentReference.networkRequest,
-        networkRequestOptions,
-      );
-      const data = readButDoNotEvaluate(
-        environment,
-        fragmentReference,
-        networkRequestOptions,
-      );
+        const readerWithRefetchQueries = readPromise(
+          fragmentReference.readerWithRefetchQueries,
+        );
 
-      const readerWithRefetchQueries = readPromise(
-        fragmentReference.readerWithRefetchQueries,
-      );
+        if (!readOutDataAndRecordsRef.current[i]) {
+          readOutDataAndRecordsRef.current[i] = {
+            records: data,
+            results: readerWithRefetchQueries.readerArtifact.resolver(
+              data.item,
+              undefined,
+            ) as ReadonlyArray<any>,
+          };
+        }
 
-      return readerWithRefetchQueries.readerArtifact.resolver(
-        data.item,
-        undefined,
-      ) as ReadonlyArray<any>;
-    });
+        return {
+          fragmentReference,
+          readerAst: readerWithRefetchQueries.readerArtifact.readerAst,
+          records: readOutDataAndRecordsRef.current[i].records,
+          callback(data) {
+            readOutDataAndRecordsRef.current[i] = {
+              records: data,
+              results: readerWithRefetchQueries.readerArtifact.resolver(
+                data.item,
+                undefined,
+              ) as ReadonlyArray<any>,
+            };
+            rerender({});
```

Comment:
This isn't ideal. I need to think about how to solve this, but we should probably use state (instead of a ref + state) because this will lead to tearing (i.e. rendering inconsistent data) in the following case:

- a user is visible in a `List` component, which renders a `UserDetail` component. In `List`, we use `useSkipLimitPagination`. `UserDetail` is an `@component`, i.e. it goes through `componentCache`, so it calls `useReadAndSubscribe`.
- `useSkipLimitPagination` uses a ref + state
- `useReadAndSubscribe` calls `setState(data)` in the callback, i.e. does not have a ref
- We're initially rendering consistent data
- We receive an update, `v1`. That updates the ref in `List` and schedules a re-render, and calls `setState(data)` in `UserDetail`
- The user then types some characters in a textbox in `List` (this tearing can occur if the user types in other components, too). React will prioritize responding to the user input more highly than processing the state changes.
- So, `List` (and `UserDetail`) are re-rendered. `List` reads from the ref, showing the updated data. `UserDetail` reads from state, so shows the stale data!

Reply:
Does this happen with `useLazyDisposableState`? I don't think so.

Fact: If a render results in a commit, no previous render will ever commit (and thus be written to the DOM). So, we only have to worry about subsequent renders. So, we can see:

- first time rendered with `cache1`, use `preCommitValue`
- commit, ref is set to `cache1`
- subsequent render with cache2, this might suspend and be part of a transition. We make a network request. Or for whatever reason, React responds to this render very slowly.
- subsequent render with cache1. This is okay, the ref still contains `cache1`.
- the render with `cache2` commits, ref is updated to contain `cache2`
- at this point, we will never see any "stale" renders with `cache1`, so we won't accidentally make any network requests.

So I think it's safe! But I could be making mistakes.